File: src/main/java/com/gitlab/pipeline/viewer/ui/TriggerPipelineDialog.java
package com.gitlab.pipeline.viewer.ui;

import com.intellij.openapi.project.Project;
import com.intellij.openapi.ui.DialogWrapper;
import com.intellij.ui.components.JBLabel;
import com.intellij.util.ui.JBUI;
import org.jetbrains.annotations.Nullable;

import javax.swing.*;
import java.awt.*;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;

/**
 * 触发流水线对话框：分支下拉选择 + 自定义变量（每行 key=value）
 */
public class TriggerPipelineDialog extends DialogWrapper {

    private final JComboBox<String> refCombo = new JComboBox<>();
    private final JTextArea variablesArea = new JTextArea(6, 40);

    public TriggerPipelineDialog(@Nullable Project project, String defaultRef, List<String> branches) {
        super(project);
        setTitle("触发流水线");
        if (branches != null) {
            for (String b : branches) {
                refCombo.addItem(b);
            }
        }
        if (refCombo.getItemCount() == 0) {
            // 分支加载失败等异常情况：回退为可手动输入，保证功能可用
            refCombo.setEditable(true);
            refCombo.getEditor().setItem(defaultRef == null ? "" : defaultRef);
        } else {
            boolean found = false;
            if (defaultRef != null) {
                for (int i = 0; i < refCombo.getItemCount(); i++) {
                    if (defaultRef.equals(refCombo.getItemAt(i))) {
                        refCombo.setSelectedIndex(i);
                        found = true;
                        break;
                    }
                }
            }
            if (!found) {
                refCombo.setSelectedIndex(0);
            }
        }
        variablesArea.setLineWrap(false);
        variablesArea.setFont(JBUI.Fonts.create("Monospaced", JBUI.scale(12)));
        init();
    }

    @Override
    protected @Nullable JComponent createCenterPanel() {
        JPanel panel = new JPanel(new GridBagLayout());
        GridBagConstraints c = new GridBagConstraints();
        c.insets = JBUI.insets(4);
        c.anchor = GridBagConstraints.WEST;

        c.gridx = 0;
        c.gridy = 0;
        panel.add(new JBLabel("分支 (ref):"), c);
        c.gridx = 1;
        c.weightx = 1;
        c.fill = GridBagConstraints.HORIZONTAL;
        panel.add(refCombo, c);

        c.gridx = 0;
        c.gridy = 1;
        c.fill = GridBagConstraints.NONE;
        c.weightx = 0;
        panel.add(new JBLabel("变量 (每行 key=value):"), c);
        c.gridx = 1;
        c.gridy = 1;
        c.fill = GridBagConstraints.BOTH;
        c.weightx = 1;
        c.weighty = 1;
        panel.add(new JScrollPane(variablesArea), c);

        panel.setPreferredSize(new Dimension(460, 240));
        return panel;
    }

    public String getRef() {
        Object item = refCombo.getSelectedItem();
        return item == null ? "" : item.toString().trim();
    }

    public Map<String, String> getVariables() {
        Map<String, String> map = new LinkedHashMap<>();
        for (String line : variablesArea.getText().split("\n")) {
            line = line.trim();
            if (line.isEmpty()) {
                continue;
            }
            int eq = line.indexOf('=');
            if (eq > 0) {
                map.put(line.substring(0, eq).trim(), line.substring(eq + 1).trim());
            } else {
                map.put(line, "");
            }
        }
        return map;
    }
}

File: src/main/kotlin/com/gitlab/pipeline/viewer/ui/TriggerPipelineDialog.kt
package com.gitlab.pipeline.viewer.ui

import com.intellij.openapi.project.Project
import com.intellij.openapi.ui.ComboBox
import com.intellij.openapi.ui.DialogWrapper
import com.intellij.ui.components.JBScrollPane
import com.intellij.ui.components.JBTextArea
import com.intellij.ui.dsl.builder.AlignX
import com.intellij.ui.dsl.builder.panel
import com.intellij.util.ui.JBUI
import org.jetbrains.annotations.Nullable

/**
 * 新建流水线对话框：分支下拉选择 + 自定义变量（每行 key=value）。
 *
 * 与原 Java 版等价语义：
 * - 分支列表非空时用下拉选择；为空时回退为可编辑下拉（用户可手动输入分支名），
 *   初始值 = 入参 defaultRef。
 * - 变量区是多行文本框，每行 `key=value`，解析时 `=` 只分割一次，空行忽略。
 * - getRef() / getVariables() 公开 API 保持原签名（外部按 String 读）。
 *
 * UI 改造点：JComboBox → [ComboBox]（主题感知下拉），JTextArea → [JBTextArea]，
 * GridBagLayout → [panel] DSL。
 */
class TriggerPipelineDialog(
    @Nullable project: Project?,
    private val defaultRef: String?,
    branches: List<String>?
) : DialogWrapper(project) {

    private val refCombo: ComboBox<String> = ComboBox()
    private val variablesArea: JBTextArea = JBTextArea().apply {
        rows = 6
        columns = 40
        lineWrap = false
        isEditable = true
        isFocusable = true
        font = JBUI.Fonts.create("Monospaced", JBUI.scale(12))
    }

    init {
        title = "新建流水线"
        val safeBranches = branches.orEmpty()
        safeBranches.forEach { refCombo.addItem(it) }
        if (refCombo.itemCount == 0) {
            // 分支加载失败等异常情况：回退为可手动输入，保证功能可用
            refCombo.isEditable = true
            refCombo.editor.item = defaultRef.orEmpty()
        } else {
            val found = defaultRef?.let { want ->
                (0 until refCombo.itemCount).firstOrNull { refCombo.getItemAt(it) == want }
            }
            if (found != null) {
                refCombo.selectedIndex = found
            } else {
                refCombo.selectedIndex = 0
            }
        }
        init()
    }

    override fun createCenterPanel() = panel {
        row("分支 (ref):") {
            cell(refCombo)
                .align(AlignX.FILL)
                .resizableColumn()
        }
        row("变量 (每行 key=value):") {
            cell(JBScrollPane(variablesArea))
                .resizableColumn()
        }
    }.apply { preferredSize = JBUI.size(480, 260) }

    /**
     * 选中的分支（去空格），不可编辑回退时直接读编辑器内容。
     */
    val ref: String
        get() = refCombo.selectedItem?.toString()?.trim().orEmpty()

    /**
     * 解析变量文本：每行 `key=value`，`=` 只切一次，无 `=` 则 value 为空串。
     * 与原 Java [java.util.LinkedHashMap] 行为一致：保留插入顺序。
     */
    val variables: Map<String, String>
        get() {
            val map = linkedMapOf<String, String>()
            for (rawLine in variablesArea.text.split("\n")) {
                val line = rawLine.trim()
                if (line.isEmpty()) continue
                val eq = line.indexOf('=')
                if (eq > 0) {
                    map[line.substring(0, eq).trim()] = line.substring(eq + 1).trim()
                } else {
                    map[line] = ""
                }
            }
            return map
        }
}
